kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/SharedKafkaProxyContext.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kubernetes.operator;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import io.javaoperatorsdk.operator.api.reconciler.Context;
+
+import io.kroxylicious.kubernetes.api.v1alpha1.KafkaProxy;
+import io.kroxylicious.kubernetes.api.v1alpha1.kafkaproxyspec.Clusters;
+import io.kroxylicious.kubernetes.operator.config.RuntimeDecl;
+
+/**
+ * Encapsulates access to the mutable state in the {@link Context<KafkaProxy>} shared between
+ * {@link ProxyReconciler} its dependent resources.
+ */
+public class SharedKafkaProxyContext {
+
+    private SharedKafkaProxyContext() {
+    }
+
+    static final String RUNTIME_DECL_KEY = "runtime";
+    static final String CLUSTER_CONDITIONS_KEY = "cluster_conditions";
+
+    /**
+     * Set the RuntimeDecl
+     */
+    static void runtimeDecl(Context<KafkaProxy> context, RuntimeDecl runtimeDecl) {
+        context.managedDependentResourceContext().put(RUNTIME_DECL_KEY, runtimeDecl);
+    }
+
+    /**
+     * Get the RuntimeDecl
+     */
+    static RuntimeDecl runtimeDecl(Context<KafkaProxy> context) {
+        return context.managedDependentResourceContext().getMandatory(RUNTIME_DECL_KEY, RuntimeDecl.class);
+    }
+
+    /**
+     * Associate a condition with a specific cluster.
+     */
+    static void addClusterCondition(Context<KafkaProxy> context, Clusters cluster, ClusterCondition clusterCondition) {
+        Map<String, ClusterCondition> map = context.managedDependentResourceContext().get(CLUSTER_CONDITIONS_KEY, Map.class).orElse(null);
```

Comment:
It's currently not a problem since we have sequential dependencies but the default `WorkflowReconcileExecutor` will execute reconciliations in parallel if it is able to. This code would be racy under parallel execution if we added further Dependants that were able to run in parallel.

Maybe we should take the result of `context.managedDependentResourceContext().put(CLUSTER_CONDITIONS_KEY, map);` and put it back into the map? Though I've just found out that `put` declares it returns T, but actually returned Optional<T>.

Reply:
PR created for the signature https://github.com/operator-framework/java-operator-sdk/pull/2587